cypress-io/cypress
# fix: relativeFile is now relative to repo root

- Closes https://github.com/cypress-io/cypress/issues/22971

### User facing changelog
The `relativeFile` property on the error `codeFrame` object is now relative to the repository root. Previously, it was relative to the Cypress config file root. 

### Additional details
- Adds internal config property `repoRoot` that uses the `.git` directory, if it exists, to identify the repository root. 
- Updates `relativeFile` in the error codeframe object to be relative to the repoRoot, if the repoRoot is set.

Why do we need this, you ask? Well, in the dashboard, we had an edge case for monorepos where we were building the incorrect URL to link to the user's source code in GitHub, BitBucket, etc. The root of the issue (see what I did there? 😛) was that the file path provided to the dashboard was relative to the cypress config. When the cypress config was located at the repo root, everything worked! But when it wasn't, as is the case for monorepos, the URL was incorrect. 

### Steps to test
1. In the `cypress` monorepo, make a failing test in `packages/app/cypress/e2e/reporter_header.cy.ts`
1.  From `cypress/packages/app`, record a run using `CYPRESS_INTERNAL_ENV=staging yarn cypress:run:e2e --headless --key ENTER_PROJECT_KEY --spec **/e2e/reporter_header.cy.ts --record --browser chrome`
1. Go to the [Dashboard staging site](https://dashboard-staging.cypress.io/). Find the test run, open the error, and click the GitHub link in the side panel. Confirm the URL works 🎉 [Here's one I made](https://dashboard-staging.cypress.io/projects/ypt4pf/runs/2648/test-results/fc170ff4-7a01-41ad-8124-89376fd5d36c)

### How has the user experience changed?
No change to user experience for standard repos. For monorepos, the `relativeFile` value is now relative to the monorepo root. See example:
**Before**: `cypress/e2e/runs.cy.ts`
**After**: `packages/app/cypress/e2e/runs.cy.ts`

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?
